package com.example.flightsearchv2.data

import android.util.Log
import androidx.datastore.core.DataStore
import androidx.datastore.preferences.core.Preferences
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.emptyPreferences
import androidx.datastore.preferences.core.stringPreferencesKey
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.flow.map
import java.io.IOException

class UserPreferencesRepository(
    private val dataStore: DataStore<Preferences>,
) {
    private companion object {
        val SEARCH_STRING = stringPreferencesKey("Search string")
        const val TAG = "UserPreferencesRepo"
    }

    // Save searching value
    suspend fun saveSearchPreferences(searchString: String) {
        dataStore.edit { preferences ->
            preferences[SEARCH_STRING] = searchString
        }
    }

    // Read value
    val searchString: Flow<String> = dataStore.data
        .catch {
            if(it is IOException) {
                Log.e(TAG, "Error reading preferences.", it)
                emit(emptyPreferences())
            } else {
                throw it
            }
        }
        .map { preferences ->
            preferences[SEARCH_STRING] ?: ""
        }
}